src/tests/functional/plugin/shared/src/behavior/ov_infer_request/memory_states.cpp

```diff
@@ -0,0 +1,286 @@
+
+    auto input = std::make_shared<ov::op::v0::Parameter>(type, shape);
+    auto mem_i1 = std::make_shared<ov::op::v0::Constant>(type, shape, 0);
+    auto mem_r1 = std::make_shared<ov::op::v3::ReadValue>(mem_i1, "r_1-3");
+    auto mul1 = std::make_shared<ov::op::v1::Multiply>(mem_r1, input);
+
+    auto mem_i2 = std::make_shared<ov::op::v0::Constant>(type, shape, 0);
+    auto mem_r2 = std::make_shared<ov::op::v3::ReadValue>(mem_i2, "c_1-3");
+    auto mul2 = std::make_shared<ov::op::v1::Multiply>(mem_r2, mul1);
+    auto mem_w2 = std::make_shared<ov::op::v3::Assign>(mul2, "c_1-3");
+
+    auto mem_w1 = std::make_shared<ov::op::v3::Assign>(mul2, "r_1-3");
+    auto sigm = std::make_shared<ov::op::v0::Sigmoid>(mul2);
+    sigm->set_friendly_name("sigmod_state");
+    sigm->get_output_tensor(0).set_names({"sigmod_state"});
+    mem_r1->set_friendly_name("Memory_1");
+    mem_r1->get_output_tensor(0).set_names({"Memory_1"});
+    mem_w1->add_control_dependency(mem_r1);
+    sigm->add_control_dependency(mem_w1);
+
+    mem_r2->set_friendly_name("Memory_2");
+    mem_r2->get_output_tensor(0).set_names({"Memory_2"});
+    mem_w2->add_control_dependency(mem_r2);
+    sigm->add_control_dependency(mem_w2);
+
+    auto function = std::make_shared<ov::Model>(ov::NodeVector{sigm}, ov::ParameterVector{input}, "add_output");
+    return function;
+}
+
+ov::CompiledModel OVInferRequestVariableStateTest::prepare_network() {
+    net->add_output("Memory_1");
+    net->add_output("Memory_2");
+    ov::Core core = createCoreWithTemplate();
+    return core.compile_model(net, deviceName, configuration);
+}
+
+TEST_P(OVInferRequestVariableStateTest, inferreq_smoke_VariableState_QueryState) {
+    auto executable_net = prepare_network();
+    auto infer_req = executable_net.create_infer_request();
+
+    auto states = infer_req.query_state();
+    ASSERT_TRUE(states.size() == 2) << "Incorrect number of VariableStates";
+
+    for (auto&& state : states) {
+        auto name = state.get_name();
+        ASSERT_TRUE(std::find(statesToQuery.begin(), statesToQuery.end(), name) != statesToQuery.end())
+            << "State " << name << "expected to be in memory states but it is not!";
+    }
+}
+
+TEST_P(OVInferRequestVariableStateTest, inferreq_smoke_VariableState_SetState) {
+    auto executable_net = prepare_network();
+    auto infer_req = executable_net.create_infer_request();
+
+    const float new_state_val = 13.0f;
+    for (auto&& state : infer_req.query_state()) {
+        state.reset();
+        auto state_val = state.get_state();
+        auto element_count = state_val.get_size();
+        std::vector<float> new_state_data;
+        for (int i = 0; i < element_count; i++) {
+            new_state_data.push_back(new_state_val);
+        }
+        ov::Tensor state_tensor = ov::Tensor(ov::element::f32, ov::Shape({1, element_count}));
+        std::memcpy(state_tensor.data(), new_state_data.data(), element_count * sizeof(float));
```

Comment:
This is test code but can be optimized to avoid temporary vector creation and multiple allocation by push_back, by fill the tensor directly like:

```suggestion
        auto state_tensor = ov::Tensor(ov::element::f32, ov::Shape({1, element_count}));
        std::fill_n(state_tensor.data<float>(), element_count, new_state_val);
```

It looks like create `state_tensor` is same in other tests, consider make local helper to avoid code duplication.